lib/tools/system-calls.js

```diff
@@ -855,42 +872,65 @@ systemCallMethods.reboot = async function reboot (retries = DEFAULT_ADB_REBOOT_R
   });
 };
 
+/**
+ * @typedef {Object} rootResult
+ * @property {boolean} isSuccessful True if the call to root/unroot was successful
+ * @property {boolean} wasAlreadyRooted True if the device was already rooted
+ */
+
 /**
  * Switch adb server to root mode.
- *
- * @return {boolean} True of the switch was successful or false
- *                   if the switch failed.
+ * @param {boolean} unroot - Should we unroot instead of root? (default false)
+ * @return {rootResult}
  */
-systemCallMethods.root = async function root () {
+systemCallMethods.changeUserPrivileges = async function changeUserPrivileges (unroot) {
+  const cmd = unroot ? 'unroot' : 'root';
+  let wasAlreadyRooted = false;
   try {
-    let {stdout} = await exec(this.executable.path, ['root']);
+    let {stdout} = await exec(this.executable.path, [cmd]);
 
     // on real devices in some situations we get an error in the stdout
-    if (stdout && stdout.indexOf('adbd cannot run as root') !== -1) {
-      throw new Error(stdout.trim());
+    if (stdout) {
+      if (stdout.includes('adbd cannot run as root')) {
+        throw new Error(stdout.trim());
+      }
+      // if the device was already rooted, return that in the result
+      if (stdout.includes('already running as root')) {
+        wasAlreadyRooted = true;
+      }
     }
-
-    return true;
+    return {isSuccessful: true, wasAlreadyRooted};
   } catch (err) {
-    log.warn(`Unable to root adb daemon: '${err.message}'. Continuing`);
-    return false;
+    const {stderr = '', message} = err;
+    log.warn(`Unable to ${cmd} adb daemon. Original error: '${message}'. Stderr: '${stderr}'. Continuing.`);
+
+    // Check the output of the stdErr to see if there's any clues that show that the device went offline
+    // and if it did go offline, restart ADB
+    const s = stderr.toLowerCase();
+    if (s.includes('closed') || s.includes('device offline')) {
+      log.warn(`Attempt to 'adb ${cmd}' caused device to go offline. Restarting adb.`);
+      await this.restartAdb();
+    }
+
+    return {isSuccessful: false, wasAlreadyRooted};
```

Comment:
`wasAlreadyRooted ` may return wrong result here. How about checking isRoot at the beginning of this method. also, we could skip the rest of the logic (and save time), if the user already had the necessary permissions